libraries/ms-common/src/stm32f0xx/adc.c

```diff
@@ -0,0 +1,139 @@
+#include <stdbool.h>
+
+#include "adc.h"
+#include "log.h"
+#include "stm32f0xx.h"
+
+typedef struct ADCInterrupt {
+  ADCCallback callback;
+  void *context;
+  uint16_t reading;
+} ADCInterrupt;
+
+typedef struct ADCStatus {
+  ADCChannel current_channel;
+  bool continuous;
+} ADCStatus;
+
+static ADCInterrupt s_adc_interrupts[NUM_ADC_CHANNEL];
+static ADCStatus s_adc_status;
+
+static void prv_adc_conversion_sequence() {
+  s_adc_status.current_channel = 0;
+  ADC_StartOfConversion(ADC1);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_EOSEQ)) {}
+}
+
+static uint32_t prv_get_channel_select() {
+  return ADC1->CHSELR;
+}
+
+void adc_init(ADCMode adc_mode) {
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_ADC1, ENABLE);
+
+  ADC_InitTypeDef adc_settings = {
+    ADC_Resolution_12b,
+    adc_mode,
+    ADC_ExternalTrigConvEdge_None,
+    ADC_ExternalTrigConv_T1_TRGO,
+    ADC_DataAlign_Right,
+    ADC_ScanDirection_Upward,
+  };
+
+  ADC_Init(ADC1, &adc_settings);
+
+  // Calculate the ADC calibration factor
+  ADC_GetCalibrationFactor(ADC1);
+
+  ADC_ContinuousModeCmd(ADC1, adc_mode);
+  while (ADC_GetFlagStatus(ADC1, ADC_FLAG_ADCAL)) { }
+
+  // Enable the ADC
+  ADC_Cmd(ADC1, ENABLE);
+  while (!ADC_GetFlagStatus(ADC1, ADC_FLAG_ADRDY)) { }
+
+  // Enable interrupts for the end of each conversion
+  stm32f0xx_interrupt_nvic_enable(ADC1_COMP_IRQn, INTERRUPT_PRIORITY_NORMAL);
+  ADC_ITConfig(ADC1, ADC_IER_EOCIE, ENABLE);
+
+  // Initialize static varables
+  s_adc_status.continuous = (ADC1->CFGR1 >> 13) & 1;
+  s_adc_status.current_channel = 0;
+}
+
+StatusCode adc_set_channel(ADCChannel adc_channel, bool new_state) {
+  if (adc_channel >= NUM_ADC_CHANNEL) {
+    return STATUS_CODE_INVALID_ARGS;
+  }
+
+  uint32_t select = prv_get_channel_select();
```

Comment:
It wasn't supposed to be used in this function. It looks like something that I forgot to remove.